x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -132,6 +135,25 @@ Listener getSearchProgressActionListener() {
         return progressListener;
     }
 
+    @Override
+    public Supplier<SearchResponseMerger> getSearchResponseMergerSupplier(
+        SearchSourceBuilder source,
+        TransportSearchAction.SearchTimeProvider timeProvider,
+        AggregationReduceContext.Builder aggReduceContextBuilder
+    ) {
+        return () -> {
+            searchResponseMerger = SearchTask.createSearchResponseMerger(source, timeProvider, aggReduceContextBuilder);
+            MutableSearchResponse mutableSearchResponse = searchResponse.get();
+            // when a remote-only CCS is done (no local shards), then onListShards is called BEFORE
+            // getSearchResponseMergerSupplier, so check if the MutableSearchResponse has been created
+            // and if so, set the SearchResponseMerger
+            if (mutableSearchResponse != null) {
+                mutableSearchResponse.setSearchResponseMerger(searchResponseMerger);
+            }
+            return searchResponseMerger;
```

Comment:
I find this method a bit hard to reason about. I have a counter proposal: how about we set the `SearchResponseMerger` to the `SearchTask` instead of setting it to the `MutableSearchResponse`. `toAsyncSearchResponse` takes the task as an argument hence the merger could be retrieved from it. This perhaps removes the need for a supplier, and for moving where the merger is created? Should we create it as we used to and set it to the task instead?

Reply:
Yes! Excellent idea. Will be in the next push.